src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/_params.py

```diff
@@ -1005,3 +1010,133 @@ def _configure_security_policy_storage_params(arg_ctx):
         c.extra('vnet_name',
                 options_list=['--vnet-name'],
                 help='The virtual network name')
+
+    ###############################################
+    #                sql managed instance         #
+    ###############################################
+    with self.argument_context('sql managed instance') as c:
+        c.argument('managed_instance_name',
+                   help='The managed instance name',
+                   options_list=['--name', '-n'],
+                   # Allow --ids command line argument. id_part=name is 1st name in uri
+                   id_part='name')
+
+    with self.argument_context('sql managed instance create') as c:
+        # Create args that will be used to build up the ManagedInstance object
+        create_args_for_complex_type(
+            c, 'parameters', ManagedInstance, [
+                'administrator_login',
+                'administrator_login_password',
+                'location',
+                'license_type',
+                'subnet_id',
+                'vcores',
+                'storage_size_in_gb',
+                'sku_name',
+            ])
+
+        c.argument('administrator_login',
+                   options_list=['--admin-user', '-u'],
+                   required=True)
+
+        c.argument('administrator_login_password',
+                   options_list=['--admin-password', '-p'],
+                   required=True)
+
+        c.argument('subnet_id',
+                   required=True,
+                   options_list=['--subnet-id', '-subn'])
+
+        c.argument('license_type',
+                   required=True,
+                   arg_type=get_enum_type(DatabaseLicenseType))
+
+        c.argument('vcores',
+                   required=True,
+                   help='Number of vcores to search for. If unspecified, all vcore sizes are shown.')
+
+        c.argument('storage_size_in_gb',
+                   required=True,
+                   help='Number of vcores to search for. If unspecified, all vcore sizes are shown.')
+
+        c.argument('sku_name',
+                   required=True)
```

Comment:
OK, will do. I have --family, --tier and --service-objective parameters now, same as for SQL db.
But, now service-objective has to be in sync with --family and --tier in order for command to work (or to have some unsupported value). For example, if I have set --family Gen4 and --tier GeneralPurpose then --service-objective has to be GP_Gen4 in order for command to work. So this is redundant parameter in that case. Is this ok behaviour? Do we have something like parameter sets in Azure CLI, similar to what we have in PowerShell?

Reply:
IMO you don't need `--service-objective`. The reason for having `--family` and `--tier` is because those are just basic strings with a small number of known values, but SKU name is a combination of those which is not obvious to new users. We have service objective for `db create` commands for backwards compatibility, I did not add --service-objective to `elastic-pool create` even though I could have.

Reply:
To answer your question, no there are no parameter sets like in PowerShell. If you look in custom.py you will see that there is code to look up the sku name in capabilities based on family and tier. It's a little complicated, but I strongly feel that it substantially improves the new user experience because the values are more discoverable.

Reply:
Also having separate params allows syntax like this: `az sql elastic-pool update -e BusinessCritical` which will update the edition but keep the family the same - very convenient for users :)

Reply:
Hmm, I am trying to remove --service-objective parameter which maps to sku.name, but this is required parameter it seems. So I would need to populate sku.name even if not requiring it in command from other two parameters (--family and --tier)? Or I am missing something?

Reply:
Yes, it needs to be populated in custom.py. See _find_elastic_pool_sku_from_capabilities for example.

Reply:
Your MI code can be mostly the same, except that:
 - MI has family capability and perf level capability at different levels so your code to find perf level will be simpler
 - All MI's have family (as opposed to DTU dbs/pools which have empty family) so you don't need allow_reset_family param.
This should reduce some code complexity.

Something like this:

```
def _find_managed_instance_sku_from_capabilities(cli_ctx, location, sku):
    '''
    Given a requested sku which may have some properties filled in
    (e.g. tier and capacity), finds the canonical matching sku
    from the given location's capabilities.
    '''

    logger.debug('_find_managed_instance_sku_from_capabilities input: %s', sku)

    if sku.name:
        # User specified sku.name, so nothing else needs to be resolved.
        logger.debug('_find_managed_instance_sku_from_capabilities return sku as is')
        return sku

    if not _any_sku_values_specified(sku):
        # User did not request any properties of sku, so just wipe it out.
        # Server side will pick a default.
        logger.debug('_find_managed_instance_sku_from_capabilities return None')
        return None

    # Some properties of sku are specified, but not name. Use the requested properties
    # to find a matching capability and copy the sku from there.

    # Get default server version capability
    capabilities_client = get_sql_capabilities_operations(cli_ctx, None)
    capabilities = capabilities_client.list_by_location(location, CapabilityGroup.supported_managed_instance_versions)
    managed_instance_version_capability = _get_default_capability(capabilities.supported_managed_instance_versions)

    # Find edition capability, based on requested sku properties
    edition_capability = _find_edition_capability(sku, server_version_capability.supported_editions)

    # Find family level capability, based on requested sku properties
    family_capability = _find_family_capability(sku, edition_capability.supported_families) # need to implement this function

    # Copy sku object from capability
    result = Sku(...)
    logger.debug('_find_managed_instance_sku_from_capabilities return: %s', result)
    return result
```

Note that db & pool capabilities return the "canonical sku" so we can copy the sku object directly from the chosen capability. You don't have this so you have to construct the sku object. This might be something Nemanja could add in managed instance capabilities.